bagel/tests/test_utility.py

```diff
@@ -543,7 +574,216 @@ def test_failed_json_reading_raises_informative_error(
+    [
+        (["20.2.7", "vA.B"], ["vA.B"]),
+        (["C.D.E", "F.G.H"], ["C.D.E", "F.G.H"]),
+    ],
+)
+def test_unrecognized_pipeline_versions_raise_error(
+    fmriprep_versions, unrecog_versions
+):
+    """Test that versions of a pipeline not found in the pipeline catalog raise an informative error."""
+    with pytest.raises(LookupError) as e:
+        dutil.check_pipeline_versions_are_recognized(
+            "fmriprep", fmriprep_versions
+        )
+
+    assert all(
+        substr in str(e.value)
+        for substr in ["unrecognized fmriprep versions"] + unrecog_versions
+    )
+
+
+def test_get_subject_imaging_sessions():
+    """Test that get_subject_imaging_sessions() accurately finds imaging sessions for a given subject."""
+    example_subject_jsonld = {
+        "identifier": "nb:34ec1e2d-9a81-4a50-bcd0-eb22c88d11e1",
+        "hasLabel": "sub-01",
+        "hasSession": [
+            {
+                "identifier": "nb:85c7473c-6122-4999-ad3b-5cd57a883c87",
+                "hasLabel": "ses-01",
+                "hasAge": 34.1,
+                "hasSex": {
+                    "identifier": "snomed:248152002",
+                    "schemaKey": "Sex",
+                },
+                "schemaKey": "PhenotypicSession",
+            },
+            {
+                "identifier": "nb:eb57d0c1-fb96-4c04-8c16-1f29f7f40db4",
+                "hasLabel": "ses-02",
+                "hasAge": 35.3,
+                "hasSex": {
+                    "identifier": "snomed:248152002",
+                    "schemaKey": "Sex",
+                },
+                "schemaKey": "PhenotypicSession",
+            },
+            {
+                "identifier": "nb:e67fd08b-9bf9-4ed8-b4cc-d0142cd27789",
+                "hasLabel": "ses-im01",
+                "hasFilePath": "/data/neurobagel/bagel-cli/bids-examples/synthetic/sub-01/ses-01",
+                "hasAcquisition": [
+                    {
+                        "identifier": "nb:5dc2e11e-4f7a-4b0e-9488-843f0a607f4b",
+                        "hasContrastType": {
+                            "identifier": "nidm:T1Weighted",
+                            "schemaKey": "Image",
+                        },
+                        "schemaKey": "Acquisition",
+                    },
+                ],
+                "schemaKey": "ImagingSession",
+            },
+        ],
+        "schemaKey": "Subject",
+    }
+    example_subject = models.Subject(**example_subject_jsonld)
+
+    assert list(
+        dutil.get_subject_imaging_sessions(example_subject).keys()
+    ) == ["ses-im01"]
+
+
+def test_create_completed_pipelines():
+    """
+    Test that completed pipelines for a subject-session are accurately identified
+    based on completion status of all pipeline steps.
+    """
+    sub_ses_data = [
+        [
+            "01",
+            "sub-01",
+            "01",
+            "ses-01",
+            "fmriprep",
+            "20.2.7",
```

Comment:
I don't get this? Why does your fmriprep have a different version across all steps? 
Are you testing that only the version tag on the last step is read out? If so, you should definitely make that explicit. 
Otherwise I would make all fmriprep versions of the same subject/session have the same version across steps. Because that's what I would expect

Reply:
>  Why does your fmriprep have a different version across all steps?

It doesn't, this example illustrates a subject-session that has run two different pipelines:
- fmriprep-20.2.7, with 2 steps (step1 - SUCCESS, step2 - FAIL)
- fmriprep-23.1.3, with 1 step (default - SUCCESS)

The goal of the test is to assert that only fmriprep-23.1.3 is interpreted by the CLI as a "completed pipeline", because **all** steps of this pipeline known to be run for the subject-session (just 1 step in this case) have a "SUCCESS" status. fmriprep-20.2.7 doesn't meet this condition because step2 failed, so it should be in the output of `create_completed_pipelines()`. I can update the docstring to be clearer.

Reply:
Aha aha, I see it now. Very cool!